pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -252,6 +255,105 @@ pub(super) async fn download_index_part(
+    }
+
+    // Stale case: If we were intentionally attached in a stale generation, there may already be a remote
+    // index in our generation.
+    //
+    // This is an optimization to avoid doing the listing for the general case below.
+    let res = do_download_index_part(storage, tenant_id, timeline_id, my_generation).await;
+    match res {
+        Ok(index_part) => {
+            tracing::debug!(
+                "Found index_part from current generation (this is a stale attachment)"
+            );
+            return Ok(index_part);
+        }
+        Err(DownloadError::NotFound) => {}
+        Err(e) => return Err(e),
+    };
+
+    // Typical case: the previous generation of this tenant was running healthily, and had uploaded
+    // and index part.  We may safely start from this index without doing a listing, because:
+    //  - We checked for current generation case above
+    //  - generations > my_generation are to be ignored
+    //  - any other indices that exist would have an older generation than `previous_gen`, and
+    //    we want to find the most recent index from a previous generation.
+    //
+    // This is an optimization to avoid doing the listing for the general case below.
+    let res =
+        do_download_index_part(storage, tenant_id, timeline_id, my_generation.previous()).await;
+    match res {
+        Ok(index_part) => {
+            tracing::debug!("Found index_part from previous generation");
+            return Ok(index_part);
+        }
+        Err(DownloadError::NotFound) => {
+            tracing::debug!(
+                "No index_part found from previous generation, falling back to listing"
+            );
+        }
+        Err(e) => {
+            return Err(e);
+        }
+    }
+
+    // General case/fallback: if there is no index at my_generation or prev_generation, then list all index_part.json
+    // objects, and select the highest one with a generation < my_generation.
+    let index_prefix = remote_index_path(tenant_id, timeline_id, Generation::none());
+    let indices = backoff::retry(
+        || async { storage.list_files(Some(&index_prefix)).await },
+        |_| false,
+        FAILED_DOWNLOAD_WARN_THRESHOLD,
+        FAILED_REMOTE_OP_RETRIES,
+        "listing index_part files",
+        // TODO: use a cancellation token (https://github.com/neondatabase/neon/issues/5066)
+        backoff::Cancel::new(CancellationToken::new(), || -> anyhow::Error {
+            unreachable!()
+        }),
+    )
+    .await
+    .map_err(DownloadError::Other)?;
+
+    // General case logic for which index to use: the latest index whose generation
+    // is <= our own.  See "Finding the remote indices for timelines" in docs/rfcs/025-generation-numbers.md
+    let max_previous_generation = indices
+        .into_iter()
+        .filter_map(parse_remote_index_path)
+        .filter(|g| g <= &my_generation)
+        .max();
```

Comment:
If `parse_remote_index_path` was returning `Result<_, _>` instead like I suggested earlier, this would have to become:

```suggestion
    let max_previous_generation = indices
        .into_iter()
        .map(|s| parse_remote_index_path(s).with_context(|| format!("parse generation of {s:?}"))))
        .filter(|g| g.is_err() || matches!(g, Ok(g) if g <= &my_generation))
        .try_fold(None, |acc, next| next.map(|next| std::cmp::max(acc, Some(next))))?;
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a7d6dc8a612031e0371d78c06bb5b5a4

Reply:
Aah but these the index_part.json names, not keys within IndexPart.

Do we store something else with the index_part.json prefix than the generations? We could. Maybe it does not make sense to return a result after all.

Reply:
The trouble with the `try_fold(...)?` approach is that it treats a parse error on an index key as an overall failure to load an index, and therefore we'll end up with a timeline in a broken state.

In some cases it's better to give up totally if we see something unexpected, but in this case I think it's better to be tolerant and keep looking for an index_part with a key we can understand.  That way:
- If a human does something like creating an index_part-0123.bak file during an incident, it doesn't break the timeline.  We would naturally prefer people didn't do that, but it can happen.
- If we later change the key format and during the rollout we have to roll back to a binary that doesn't understand the new format, it is desirable for the old version to just use whichever index_part object had a key that it did understand, rather than going into a broken state.

Reply:
Yes makes sense. Did not see that originally.